s3_storage_provider.py

```diff
@@ -132,35 +144,34 @@ class _S3DownloadThread(threading.Thread):
         deferred (Deferred[_S3Responder|None]): If file exists
             resolved with an _S3Responder instance, if it doesn't
             exist then resolves with None.
+        parent_logcontext (LoggingContext): the logcontext to report logs and metrics
+            against.
     """
+    with LoggingContext(parent_context=parent_logcontext):
+        logger.info("Fetching %s from S3", key)
 
-    def __init__(self, bucket, api_kwargs, key, deferred):
-        super(_S3DownloadThread, self).__init__(name="s3-download")
-        self.bucket = bucket
-        self.api_kwargs = api_kwargs
-        self.key = key
-        self.deferred = deferred
-
-    def run(self):
         local_data = threading.local()
-        if not hasattr(local_data, "b3_session"):
-            local_data.b3_session = boto3.session.Session()
-        session = local_data.b3_session
-        s3 = session.client('s3', **self.api_kwargs)
 
         try:
-            resp = s3.get_object(Bucket=self.bucket, Key=self.key)
+            s3 = local_data.b3_client
+        except AttributeError:
+            b3_session = boto3.session.Session()
+            local_data.b3_client = s3 = b3_session.client("s3", **api_kwargs)
+
+        try:
+            resp = s3.get_object(Bucket=bucket, Key=key)
         except botocore.exceptions.ClientError as e:
-            if e.response['Error']['Code'] in ("404", "NoSuchKey",):
-                reactor.callFromThread(self.deferred.callback, None)
+            if e.response["Error"]["Code"] in ("404", "NoSuchKey",):
+                logger.info("Media %s not found in S3", key)
+                reactor.callFromThread(deferred.callback, None)
                 return
 
-            reactor.callFromThread(self.deferred.errback, Failure())
+            reactor.callFromThread(deferred.errback, Failure())
             return
 
         producer = _S3Responder()
-        reactor.callFromThread(self.deferred.callback, producer)
-        _stream_to_producer(reactor, producer, resp["Body"], timeout=90.)
+        reactor.callFromThread(deferred.callback, producer)
+        _stream_to_producer(reactor, producer, resp["Body"], timeout=90.0)
```

Comment:
Also wondering if this would benefit from being configurable (eg, for maximum sized media files, 90s may or may not be long enough)

Reply:
possibly, but that would be a new feature for existing code, and I'm trying not to get distracted into fixing everything that could be possibly be fixed for this thing.